Readme.pod
=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Search - Search Engines for Mojolicious


=head1 SYNOPSIS

  use Mojolicious::Lite;

  # Load plugin in Mojolicious
  plugin Search => {
    items_per_page => 25,

    # Engine specific configuration:
    engine => 'Lucy',
    schema => {
      content => {
        type => 'fulltext',
        highlightable => 1,
        stored => 1
      },
      title => {
        type => 'fulltext',
        highlightable => 1,
        stored => 1
      },
      url => {
        type => 'string',
        indexed => 0
      }
    },

    # Initialize index with some documents
    on_init => sub {
      shift->add(
        {
          title => 'My first Ballon',
	  content => 'Please find me!',
          url => '/page1'
        },
        {
          title => 'My second Ballon',
	  content => 'Thanks for searching',
	  url => '/page2'
        }
      );
    }
  };

  get '/' => sub {
    shift->render('index');
  };

  app->start;

  __DATA__

  @@ layouts/default.html.ep
  <!DOCTYPE html>
  <html>
    <head><title><%= title %></title></head>
    <body>
      <h1><%= title %></h1>
  %= content
    </body>
  </html>

  @@ index.html.ep
  % layout 'default', title => 'Search';

  %# Set search form
  %= form_for url_for, begin
  %=   search_field 'q'
  % end

  %= search query => param('q'), begin
  % title 'Search for ' . search->query;

  <p>Found <%= search->total_results %> matches</p>
  <ul>
  %=  search_results begin
    <li>
      <h2><%= link_to $_->{title}, $_->{url} %></h2>
      <p><%= $_->{content} %></p>
    </li>
  %   end
  </ul>
  % end


=head1 DESCRIPTION

L<Mojolicious::Plugin::Search> is a base plugin to add
search capabilities to Mojolicious.

B<This is early software, please use it with care! Things may change until it is on CPAN!>


=head1 METHODS

L<Mojolicious::Plugin::Search> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.


=head2 register

Called when registering the plugin.
All parameters can be set either on registration or as part
of the configuration file with the key C<Search>
(with the configuration file having the higher precedence).


Accepts the following parameters:

=over 2

=item B<items_per_page>

Set the maximum number of hits shown per page.


=item B<engine>

A L<Mojolicious::Plugin::Search> engine.
Defaults to
L<Lucy|Mojolicious::Plugin::Search::Lucy>. See
L<Mojolicious::Plugin::Search::Lucy> for further information.

=back

All further parameters are passed to the chosen engine.


=head1 HELPERS

=head2 search

  # In controller (blocking)
  my $result = $c->search(q => 'test');
  # print $result->total_results;
  # print $result->hit(2)->title;

  # In controller (non-blocking)
  $c->search(
    query => 'name',
    start_page => 4,
    count => 25,
    cb => sub {
      my $index = shift;
      $c->render(text => 'Found ' . $index->total_results . ' matches!');
    }
  );

  # In Templates (blocking)
  %= search query => 'test', begin
  %# ...
  % end

Initializes a search.

The parameter C<query> defines the query terms,
C<start_page> defines the numerical starting page of the search,
C<count> defines the number of results that should be shown per page.
If C<count> is greater than C<items_per_page>, it will be limited to that.

The parameter C<cb> can be used to pass a callback method to be evaluated
(e.g., in case of a non-blocking search). The parameter C<template> may define
a template to render after a search (also for non-blocking searches).

C<query> and C<start_page> may also be defined in advanced
using the L<Mojolicious::Controller/stash|stash>, if prefixed with C<search.>.
C<count> may also be set as C<search.items_per_page>, overriding
the plugin default value limit.

In addition to theses parameters, L</search|search> accepts arbitrary engine
specific parameters.

When called without C<cb> or C<template>, the method returns a
L<Mojolicious::Plugin::Search::Index|Index object>,
representing search parameters and probably results.
When called with C<cb> the callback is executed with the
L<Mojolicious::Plugin::Search::Index|Index object> as the first parameter.

In templates, L</search|search> can be used either as a tag helper, expecting a
nested L</search_results|search_results> block for displaying the results
(This usage is, however, not recommended for long-term searches!),
or as a proxy object for accessing L<Mojolicious::Plugin::Search::Index|Index object>.

  %# In templates
  <p>Found <%= search->total_results %> matches</p>
  <p>Show page <%= search->start_page %>/<%= search->total_pages %></p>


=head2 search_results

  %= search begin
  %=  search_results begin
    <p><%= $_->{title} %></p>
    <p><%= $_->{content} %></p>
  %   end
  % end

Iterates over all hits for a certain query, contains a hit object
in C<$_> and the stash value C<search.hit>.
The concrete realization of the hit is engine specific.


=head1 ENGINES

L<Mojolicious::Plugin::Search> is bundled with an
L<Apache Lucy|Mojolicious::Plugin::Search::Lucy> engine.


=head1 DEPENDENCIES

L<Mojolicious> (best with SSL support),
L<Lucy>,
L<Mojolicious::Plugin::Util::Endpoint>.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Search


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut


Changes
0.08 2018-10-09
        - Fixed test suite for Mojolicious > 7.79.
          (Was broken before, but accepted until Mojo::DOM::CSS
          became more correct)

0.07 2017-04-24
        - Updated to recent version of Mojolicious.
        - Fixed minor issue for unset items_per_page value.

0.06 2016-12-01
        - Updated Lucy dependency.

0.05 2015-01-16
        - Added lucy command.

0.04 2014-12-01
        - Pass engine_class to register for easy adding
	  index attributes.

0.03 2014-11-30
        - Pretty much a rewrite to support non-blocking
	  and limit use of the stash.
	- Decoupling of OpenSearch.

0.02 2014-06-08
        - No more transparent support of query parameters.

0.01 2014-05-20
        - Initial submission to GitHub.
        - Initial submission to CPAN.


t/Mojolicious/Plugin/Search/Example.pm
package Mojolicious::Plugin::Search::Example;
use Mojo::Base 'Mojolicious::Plugin';

# Contain documents
our @DOCS = ();

# Register plugin
sub register {
  my ($self, $mojo, $index_class, $param) = @_;

  # Add attributes to the index class
  $index_class->attr(index_name => 'My Example engine');

  # Initialize some documents
  @DOCS = (
    {
      title => 'Test',
      content => 'My first example'
    },
    {
      title => 'Example!',
      content => 'My second example'
    }
  );
};

# Search method for documents
sub search {
  my ($self, $index) = @_;
  my $cb = pop
    if $_[-1] && ref $_[-1] && ref $_[-1] eq 'CODE';

  # Get parameters from index object
  my $query = $index->query;
  my $count = $index->items_per_page;
  my $start_index = $index->start_index;

  # Collect hits
  my @matches = ();
  my $total_results = 0;

  # Iterate sequentially through all docs
  foreach (grep { $_ } @DOCS) {

    # Check if the doc is a match
    if (index($_->{title}.' '.$_->{content}, $query) >= 0) {

      # Check if match is in scope
      if ($total_results >= $start_index &&
	    $total_results < ($start_index + $count)) {
	push(@matches, $_);
      };

      # Match!
      $total_results++;
    };
  };

  # Set overall matches
  $index->total_results($total_results);

  # Set result array
  $index->results(@matches);

  # Execute callback in case of non-blocking like request
  return $cb->($index) if $cb;
  return 1;
};


# Add new document
sub add {
  my ($self, $index, $obj) = @_;
  push(@DOCS, $obj);
  return 1;
};


# Delete document from index
sub delete {
  my ($self, $index, $i) = @_;
  $DOCS[$i] = undef;
  return 1;
};


1;


t/Mojolicious/Plugin/Search/Failing.pm
package Mojolicious::Plugin::Search::Failing;
use Mojo::Base 'Mojolicious::Plugin';

# Engine missing a search method
sub register {};

1;


Makefile.PL
#!/usr/bin/env perl
use 5.010001;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME         => 'Mojolicious::Plugin::Search',
  AUTHOR       => 'Nils Diewald',
  ABSTRACT     => 'Search Engines for Mojolicious',
  VERSION_FROM => 'lib/Mojolicious/Plugin/Search.pm',
  BUILD_REQUIRES => {
    'Test::More' => 0,
    'File::Temp' => 0
  },
  PREREQ_PM => {
    'Mojolicious' => 7.15,
    'Mojolicious::Plugin::Util::Endpoint' => 0.19,
    'Lucy' => '0.006001'
  },
  LICENSE      => 'artistic_2',
  META_MERGE   => {
    requires  => { perl => '5.010001' },
    resources => {
      license     => 'http://www.opensource.org/licenses/artistic-license-2.0',
      repository  => 'https://github.com/Akron/Mojolicious-Plugin-Search'
    }
  },
  MIN_PERL_VERSION => '5.010001',
  test => {TESTS => 't/*.t'}
);


lib/Mojolicious/Plugin/Search/Index.pm
package Mojolicious::Plugin::Search::Index;
use Mojo::Base -base;
use Mojo::Collection;
use Scalar::Util 'blessed';
use POSIX 'ceil';

has [qw/controller engine query items_per_page/];
has start_page => 1;
has total_results => -1;

# Get total pages
sub total_pages {
  my $self = shift;
  return 0 if $self->total_results <= 0;
  return ceil($self->total_results / ($self->items_per_page || 1));
};


# Get or set results
sub results {
  my $self = shift;
  if (@_ > 0) {
    return $self->{results} = Mojo::Collection->new(@_);
  };
  return $self->{results} // Mojo::Collection->new;
};


# Get or set wanted start index
sub start_index {
  my $self = shift;
  if ($_[0]) {
    $self->{start_index} = $_[0];
    return;
  };

  return $self->{start_index} //
    ($self->start_page - 1) * ($self->items_per_page || 1);
};


# Get hit
sub hit {
  return $_[0]->results->[$_[1] // 0];
};


# Call autoload
sub AUTOLOAD {
  my $self = shift;

  # Get some information
  my ($package, $method) = our $AUTOLOAD =~ /^(.+)::(.+)$/;

  # Call method on engine
  my $engine = $self->engine;
  if ($engine->can($method)) {
    return $engine->$method($self, @_);
  };

  # Croak on problem
  Carp::croak qq{Can't locate object method "$method" via package } .
      blessed $engine;
};


1;


__END__

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Search::Index - Index object for Search


=head1 SYNOPSIS

  use Mojolicious::Plugin::Search::Index;

  my $index = Mojolicious::Plugin::Search::Index->new(
    controller => Mojolicious::Controller->new(...),
    engine     => $engine
  );

  $index->query("tree AND forest");
  $index->total_results(7);

=head1 DESCRIPTION

L<Mojolicious::Plugin::Search::Index> is a proxy object for search engines,
used by L<Mojolicious::Plugin::Search>.

=head1 ATTRIBUTES

=head2 controller

  $index->controller(Mojolicious::Controller->new);
  print $index->controller->stash('search.hit');

Get or set the controller object.
Normally set by L<Mojolicious::Plugin::Search>.


=head2 engine

  $index->controller(Mojolicious::Plugin::Search::Lucy->new);
  print $index->controller->stash('search.hit');

Get or set the controller object.
Normally set by L<Mojolicious::Plugin::Search>.


=head2 hit

  print $index->hit(0)->{title};

Get a hit by providing an index integer.
Only hits in the scope of search are accessible.
Normally set by a L<search|Mojolicious::Plugin::Search/search>.


=head2 items_per_page

  $index->items_per_page(20);
  print $index->items_per_page;

Get or set the number of items to show per page.
Defaults to the value of the calling
L<Mojolicious::Plugin::Search/items_per_page|plugin>.


=head2 query

  $index->query("tree OR forest);
  print $index->query;

Get or set the query term of the search.

=head2 results

  $index->results({
    title => 'Foo'
  },{
    title => 'Bar'
  });

  say $index->results->map(sub { $_->{title} })->join(', ');

Set search results as an array or return search results as
a L<Mojo::Collection>.


=head2 start_index

  $index->start_index(3);
  print $index->start_index;

Get or set the start index for offsets of search results.
Defaults to the calculated value based on L<start_page|/start_page>
and L<items_per_page|/items_per_page>.


=head2 start_page

  $index->start_page(2);
  print $index->start_page;

Get or set the start page for offsets of search results.
Defaults to C<1>.


=head2 total_pages

  print $index->total_pages;

Get the number of pages of results based on the values of
L<total_results|/total_results> and L<items_per_page|/items_per_page>.


=head2 total_results

  $index->total_results(3);
  print $index->total_results;

Get or set the number of total_results of a search.
Normally set by a L<search|Mojolicious::Plugin::Search/search>.


=head1 AUTOLOAD

In addition to the L<ATTRIBUTES|/ATTRIBUTES> above, the object autoloads
further methods provided by the associated engine. Methods called on the
object but provided by the engine will get the C<index> object passed as
the second argument after the invocant and followed by further parameters.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Search


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut



lib/Mojolicious/Plugin/Search/Lucy/lucy.pm
package Mojolicious::Plugin::Search::Lucy::lucy;
use Mojo::Base 'Mojolicious::Command';

use Getopt::Long qw/GetOptions :config no_auto_abbrev no_ignore_case/;

has description => 'Manage your Lucy index.';
has usage       => sub { shift->extract_usage };

# Run chi
sub run {
  my $self = shift;

  my $command = shift;

  print $self->usage and return unless $command;

  # Get the application
  my $app = $self->app;
  my $log = $app->log;

  # List all associated caches
  if ($command eq 'optimize') {

    my $i = $app->search->new_indexer;

    if ($i) {
      $i->optimize;
      print "Index successfully optimized!\n";
      return 1;
    }
    else {
      print "Unable to open index!\n";
    };
  };

  # Unknown command
  print $self->usage and return;
};


1;


__END__

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Lucy::lucy - Manage your Lucy Index


=head1 SYNOPSIS

  usage: perl app.pl lucy <command>

    perl app.pl lucy optimize

  Interact with Lucy indices associated with your application.
  Valid commands include:

    optimize
      Optimize your index for search performance.
      Warning: This may take some time!


=head1 DESCRIPTION

L<Mojolicious::Plugin::Search::Lucy::lucy> helps you to manage your
Lucy index associated with L<Mojolicious::Plugin::Search>.


=head1 ATTRIBUTES

L<Mojolicious::Plugin::Search::Lucy::lucy> inherits all attributes
from L<Mojolicious::Command> and implements the following new ones.


=head2 description

  my $description = $lucy->description;
  $lucy = $lucy->description('Foo!');

Short description of this command, used for the command list.


=head2 usage

  my $usage = $lucy->usage;
  $lucy = $lucy->usage('Foo!');

Usage information for this command, used for the help screen.


=head1 METHODS

L<Mojolicious::Plugin::Search::Lucy::lucy> inherits all methods from
L<Mojolicious::Command> and implements the following new ones.


=head2 run

  $lucy->run;

Run this command.


=head1 DEPENDENCIES

L<Mojolicious>,
L<Lucy>.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Search


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015, L<Nils Diewald||http://nils-diewald.de>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

The documentation is based on L<Mojolicious::Command::eval>,
written by Sebastian Riedel.

=cut


lib/Mojolicious/Plugin/Search/OpenSearch.pm
package Mojolicious::Plugin::Search::OpenSearch;
use Mojo::Base 'Mojolicious::Plugin';
use File::Basename 'dirname';
use File::Spec::Functions 'catdir';

# Todo: Support json!

our @OS_PARAM_DESC = qw/short_name description long_name
			tags contact developer attribution
			syndication_right language/;

# Register plugin
sub register {
  my ($plugin, $mojo, $param) = @_;

  # Check that search is loaded in advance

  # Load parameter from Config file
  if (my $config_param = $mojo->config('Search-OpenSearch')) {
    $param = { %$config_param, %$param };
  };

  # Add templates
  push @{$mojo->renderer->paths}, catdir(dirname(__FILE__), 'OpenSearch','templates');

  my $helpers = $mojo->renderer->helpers;

  # Get endpoint plugin
  unless (exists $helpers->{endpoint}) {
    $mojo->plugin('Util::Endpoint');
  };

  # Missing information
  unless ($param->{short_name} && $param->{description}) {
    $mojo->log->warn('Missing "short_name" or "description"');
  };

  # Use moniker as short name
  $param->{short_name} //= $mojo->moniker;

  # Simple description
  $param->{description} //= 'OpenSearch';

  # Establish 'opensearch' shortcut
  $mojo->routes->add_shortcut(
    opensearch => sub {
      my $r = shift;
      my %osparam = @_;

      my $hit = delete $osparam{hit};

      if (!$hit || ref $hit ne 'CODE') {
	$mojo->log->error('No hit callback established');
	return;
      };

      my @osattr = qw/count startPage startIndex language/;

      my %hash;
      # Mandatory parameters
      foreach (qw/format searchTerms/) {
	$osparam{$_} //= $_;
	$hash{ $osparam{$_} } = '{' . $_ . '}';
      };

      # Optional parameters
      foreach (@osattr) {
	# qw/inputEncoding outputEncoding/
	$osparam{$_} //= $_ ;
	$hash{ $osparam{$_} } = '{' . $_ . '?}';
      };

      # Establish endpoint
      $r->endpoint(opensearch => { query => [ %hash ] });

      # Establish callback
      $r->any->to(
	cb => sub {
	  my $c = shift;

	  my $searchTerms = $c->param( $osparam{searchTerms} );

	  # No searchTerms defined
	  unless ($searchTerms) {
	    return $c->render_not_found(
	      text => 'The request is missing a search term',
	      status => '400' # Bad request
	    );
	  };

	  # Set opensearch stash values
	  foreach (@OS_PARAM_DESC) {
	    $c->stash('opensearch.' . $_ => $param->{$_}) if $param->{$_};
	  };

	  # Start creating query
	  $c->stash('search.searchTerms' => $searchTerms);

	  # Count, StartPage, startIndex, language
	  foreach (@osattr) {
	    my $key = $osparam{$_};
	    my $val = $c->param($key) or next;
	    if ($_ ne 'language') {

	      # check all numbers are numbers
	      next if $val !~ s/^\s*(\d+)\s*/$1/;
	    };
	    $c->stash('search.' . $key => $val);
	  };

	  # Set default format
	  $c->stash('format' => scalar($c->param('format')) // 'rss');

	  $c->render(
	    template => 'opensearch/results',
	    hit => $hit
	  );
	}
      );
    }
  );

  # Description file
  my $desc = delete $param->{file} || '/.well-known/opensearch.xml';

  $mojo->routes->get($desc)->to(
    cb => sub {
      my $c = shift;

      # Set opensearch stash values
      foreach (@OS_PARAM_DESC) {
	$c->stash('opensearch.' . $_ => $param->{$_}) if $param->{$_};
      };

      # Set content type
      $c->res->headers->content_type('application/opensearchdescription+xml');

      # TODO: Render this in advance
      return $c->render(
	template => 'opensearch/description',
	format   => 'xml'
      );
    })->name('opensearch-description');
};


1;


__END__

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Search::OpenSearch - OpenSearch for Mojolicious


=head1 SYNOPSIS


  use Mojolicious::Lite;

  plugin Search => {};

  plugin 'Search::OpenSearch' => {
    short_name => 'My search',
    description => 'Search my site using OpenSearch'
  };

  get('/search/open')->opensearch(
    hit => sub {
      my ($c, $hit) = @_;
      return {
        title => $hit->{title},
        link => $hit->{url},
        snippet => $hit->{highlight}
      }
    }
  );


=head1 DESCRIPTION

L<Mojolicious::Plugin::Search::OpenSearch> is an L<OpenSearch|http://www.opensearch.org>
extension to L<Mojolicious::Plugin::Search>.

B<This is early software, please use it with care.>

=head1 METHODS

L<Mojolicious::Plugin::Search::OpenSearch> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.


=head2 register

Called when registering the plugin.
All parameters can be set either on registration or as part
of the configuration file with the key C<Search-OpenSearch>.

Accepts the parameters C<short_name>, C<description>, C<language>,
C<long_name>, C<developer>, C<contact>, C<adult_content>, C<attribution>,
and C<syndication_right>.
See the L<specification|http://www.opensearch.org/Specifications/OpenSearch/1.1#OpenSearch_description_document> for further information on these attributes.
Both C<short_name> and C<description> are mandatory.

Additionally supports the attribute C<file>, setting the path to the
description document. Defaults to C</.well-known/opensearch.xml>.


=head1 SHORTCUTS

=head2 opensearch

  get('/myopensearch')->opensearch(
    searchTerms => 'q',
    startPage => 'page',
    hit => sub {
      my ($c, $hit) = @_;
      return {
        title => $hit->{title},
        link => $hit->{url},
        snippet => $hit->{highlight}
      }
    }
  );

Defines the endpoint for C<rss> and C<atom> responses.
Accepts parameters to rewrite the name of the following
query parameters supported by OpenSearch: C<count>, C<startPage>,
C<startIndex>, C<language> and C<searchTerms>.
See the L<specification|http://www.opensearch.org/Specifications/OpenSearch/1.1#OpenSearch_URL_template_syntax> for further information on these attributes.

=over 2

=item B<hit>

In adition to the parameters mentioned above, the C<hit> parameter expects a
callback function, that is called for each hit.
Passed parameters is the controller object and the hit object.
Expects a hash reference with string values for C<title>, C<link> and C<snippet>
of the hit.

=back

Establishes an L<endpoint|Mojolicious::Plugin::Util::Endpoint> with
the name C<opensearch>.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Search


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut


lib/Mojolicious/Plugin/Search.pm
package Mojolicious::Plugin::Search;
use Mojo::Base 'Mojolicious::Plugin';
use Mojolicious::Plugin::Search::Index;
use Scalar::Util 'looks_like_number';
use Mojo::Util qw/camelize/;
use Mojo::ByteStream 'b';

our $VERSION = '0.08';

# TODO: Improve the documentation by using similar
#       attribute style as in M::P::S::Lucy
# TODO: Check for failing engines (e.g. methods missing)

# Maximum number of items per page
has items_per_page => 25;
has engine => __PACKAGE__ . '::Lucy';


# Register plugin
sub register {
  my ($plugin, $app, $param) = @_;
  $param ||= {};

  # Load parameter from Config file
  if (my $config_param = $app->config('Search')) {
    $param = { %$param, %$config_param };
  };

  # Set engine
  $plugin->engine(delete $param->{engine})
    if $param->{engine};

  # Set items per page if given
  $plugin->items_per_page(delete $param->{items_per_page})
    if $param->{items_per_page};

  # Get specific search engine
  my $engine = camelize($plugin->engine);

  # Engine is relative
  $engine = __PACKAGE__ . '::' . $engine if index($engine, '::') < 0;

  # Load and register the engine
  my $e = $app->plugins->load_plugin($engine) or return;
  $e->register($app, __PACKAGE__ . '::Index', $param);

  # No search method in plugin
  unless ($e->can('search')) {
    $app->log->error(qq!$engine expects a 'search' method!);
    return;
  };

  # Establishes 'search' helper
  $app->helper(
    search => sub {
      my $c = shift;

      # Return searcher object
      if (@_ == 0) {

	# 'Index' object may already be initialized
	return $c->stash->{'search._index'} //= _index(
	  controller => $c,
	  engine => $e
	);
      };

      # Get new 'Index' object
      my $index = _index(
	controller => $c,
	engine => $e
      );
      $c->stash->{'search._index'} = $index;

      # Get callback for templates
      my $cb = pop if (@_ % 2) != 0 && ref $_[-1] && ref $_[-1] eq 'CODE';

      # Get parameters for search
      my %param = @_;

      # Set query
      $index->query(delete $param{query} // $c->stash('search.query'));

      # No query defined
      return unless $index->query;

      # Set count
      if ($param{count} &&
	    looks_like_number($param{count}) &&
	      $param{count} <= $plugin->items_per_page ) {
	$index->items_per_page(delete $param{count});
      }
      # already set by stash - or use plugin param
      else {
	$index->items_per_page(
	  $c->stash('search.count') // $plugin->items_per_page
	);
      };

      # Set start page based on param
      if ($param{start_page} && looks_like_number($param{start_page})) {
	$index->start_page(delete $param{start_page});
      }
      # already set by stash
      elsif ($c->stash('search.start_page')) {
	$index->start_page($c->stash('search.start_page'));
      };

      # Non-blocking call (at least possible for engines)
      if ($param{cb} || $param{template}) {

	# Prevent automatic rendering
	$c->render_later;

	# Search non-blocking with callback
	if (my $cb = delete $param{cb}) {
	  $e->search($index, @_, $cb);
	}

	# Search non-blocking with template information
	elsif (my $template = delete $param{template}) {
	  $e->search(
	    $index, @_, sub {
	      $c->render(template => $template);
	    });
	};

	return 1;
      };

      # Blocking call
      $e->search($index, @_);

      # Return object in case it's not called in a template
      return $index unless $cb;

      # Template callback
      my $value = $cb->($c);

      # Remove hits from stash (maybe not necessary)
      foreach (qw/results total_results _index/) {
	delete $c->stash->{"search.$_"};
      };

      # Return value
      return $value;
    }
  );


  # Establish 'search_results' taghelper
  $app->helper(
    search_results => sub {
      my $c = shift;

      # Get 'Index' object from stash
      my $index = $c->stash('search._index') or return;

      # This is a tag helper for templates
      my $cb = shift;
      if (!ref $cb || !(ref $cb eq 'CODE')) {
	$c->app->log->error('search_results expects a code block');
	return '';
      };

      # Iterate over results
      my $string = $index->results->map(
	sub {
	  # Call hit callback
	  $c->stash('search.hit' => $_);
	  local $_ = $_[0];
	  return $cb->($_);
	})->join;

      # Remove hit from stash
      delete $c->stash->{'search.hit'};
      return b($string);
    }
  );


  # Run init on engine (that's not really recommended)
  if ($param->{on_init} &&
	ref $param->{on_init} &&
	  ref $param->{on_init} eq 'CODE') {
    $param->{on_init}->(
      _index(
	controller => $app->build_controller,
	engine => $e
      )
    );
  };
};


# Return 'Index' object
sub _index {
  Mojolicious::Plugin::Search::Index->new(@_);
};


1;


__END__

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Search - Search Engines for Mojolicious


=head1 SYNOPSIS

  use Mojolicious::Lite;

  # Load plugin in Mojolicious
  plugin Search => {
    items_per_page => 25,

    # Engine specific configuration:
    engine => 'Lucy',
    schema => {
      content => {
        type => 'fulltext',
        highlightable => 1,
        stored => 1
      },
      title => {
        type => 'fulltext',
        highlightable => 1,
        stored => 1
      },
      url => {
        type => 'string',
        indexed => 0
      }
    },

    # Initialize index with some documents
    on_init => sub {
      shift->add(
        {
          title => 'My first Ballon',
	  content => 'Please find me!',
          url => '/page1'
        },
        {
          title => 'My second Ballon',
	  content => 'Thanks for searching',
	  url => '/page2'
        }
      );
    }
  };

  get '/' => sub {
    shift->render('index');
  };

  app->start;

  __DATA__

  @@ layouts/default.html.ep
  <!DOCTYPE html>
  <html>
    <head><title><%= title %></title></head>
    <body>
      <h1><%= title %></h1>
  %= content
    </body>
  </html>

  @@ index.html.ep
  % layout 'default', title => 'Search';

  %# Set search form
  %= form_for url_for, begin
  %=   search_field 'q'
  % end

  %= search query => param('q'), begin
  % title 'Search for ' . search->query;

  <p>Found <%= search->total_results %> matches</p>
  <ul>
  %=  search_results begin
    <li>
      <h2><%= link_to $_->{title}, $_->{url} %></h2>
      <p><%= $_->{content} %></p>
    </li>
  %   end
  </ul>
  % end


=head1 DESCRIPTION

L<Mojolicious::Plugin::Search> is a base plugin to add
search capabilities to Mojolicious.

B<This is early software, please use it with care! Things may change until it is on CPAN! It is also likely to be renamed to be Lucy-specific again.>


=head1 METHODS

L<Mojolicious::Plugin::Search> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.


=head2 register

Called when registering the plugin.
All parameters can be set either on registration or as part
of the configuration file with the key C<Search>
(with the configuration file having the higher precedence).


Accepts the following parameters:

=over 2

=item B<items_per_page>

Set the maximum number of hits shown per page.


=item B<engine>

A L<Mojolicious::Plugin::Search> engine.
Defaults to
L<Lucy|Mojolicious::Plugin::Search::Lucy>. See
L<Mojolicious::Plugin::Search::Lucy> for further information.

=back

All further parameters are passed to the chosen engine.


=head1 HELPERS

=head2 search

  # In controller (blocking)
  my $result = $c->search(q => 'test');
  # print $result->total_results;
  # print $result->hit(2)->title;

  # In controller (non-blocking)
  $c->search(
    query => 'name',
    start_page => 4,
    count => 25,
    cb => sub {
      my $index = shift;
      $c->render(text => 'Found ' . $index->total_results . ' matches!');
    }
  );

  # In Templates (blocking)
  %= search query => 'test', begin
  %# ...
  % end

Initializes a search.

The parameter C<query> defines the query terms,
C<start_page> defines the numerical starting page of the search,
C<count> defines the number of results that should be shown per page.
If C<count> is greater than C<items_per_page>, it will be limited to that.

The parameter C<cb> can be used to pass a callback method to be evaluated
(e.g., in case of a non-blocking search). The parameter C<template> may define
a template to render after a search (also for non-blocking searches).

C<query> and C<start_page> may also be defined in advanced
using the L<Mojolicious::Controller/stash|stash>, if prefixed with C<search.>.
C<count> may also be set as C<search.items_per_page>, overriding
the plugin default value limit.

In addition to theses parameters, L</search|search> accepts arbitrary engine
specific parameters.

When called without C<cb> or C<template>, the method returns a
L<Mojolicious::Plugin::Search::Index|Index object>,
representing search parameters and probably results.
When called with C<cb> the callback is executed with the
L<Mojolicious::Plugin::Search::Index|Index object> as the first parameter.

In templates, L</search|search> can be used either as a tag helper, expecting a
nested L</search_results|search_results> block for displaying the results
(This usage is, however, not recommended for long-term searches!),
or as a proxy object for accessing L<Mojolicious::Plugin::Search::Index|Index object>.

  %# In templates
  <p>Found <%= search->total_results %> matches</p>
  <p>Show page <%= search->start_page %>/<%= search->total_pages %></p>


=head2 search_results

  %= search begin
  %=  search_results begin
    <p><%= $_->{title} %></p>
    <p><%= $_->{content} %></p>
  %   end
  % end

Iterates over all hits for a certain query, contains a hit object
in C<$_> and the stash value C<search.hit>.
The concrete realization of the hit is engine specific.


=head1 ENGINES

L<Mojolicious::Plugin::Search> is bundled with an
L<Apache Lucy|Mojolicious::Plugin::Search::Lucy> engine.


=head1 DEPENDENCIES

L<Mojolicious> (best with SSL support),
L<Lucy>,
L<Mojolicious::Plugin::Util::Endpoint>.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Search


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2018, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut
